pygeos/testing.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+from numpy.testing import assert_almost_equal
+
+import pygeos
+
+__all__ = ["assert_geometries_equal"]
+
+
+def _assert_nan_coords_same(x, y, decimal):
+    x_ndim = pygeos.get_coordinate_dimension(x)
+    y_ndim = pygeos.get_coordinate_dimension(y)
+    if x_ndim != y_ndim:
+        return False
+
+    x_coords = pygeos.get_coordinates(x, include_z=x_ndim == 3)
+    y_coords = pygeos.get_coordinates(y, include_z=x_ndim == 3)
+
+    # Check NaN equality
+    try:
+        assert_almost_equal(x_coords, y_coords)
+    except AssertionError:
+        return False
+    else:
+        return True
+
+
+def _assert_none_same(x, y):
+    x_id = pygeos.is_missing(x)
+    y_id = pygeos.is_missing(y)
+
+    if not (x_id == y_id).all():
+        raise AssertionError(
+            "One of the arrays contains a None where the other has a geometry."
+        )
+
+    # If there is a scalar, then here we know the array has the same
+    # flag as it everywhere, so we should return the scalar flag.
+    if isinstance(x_id, bool) or x_id.ndim == 0:
+        return bool(x_id)
+    elif isinstance(y_id, bool) or y_id.ndim == 0:
+        return bool(y_id)
+    else:
+        return y_id
+
+
+def assert_geometries_equal(x, y, decimal=7, equal_none=True, equal_nan=True):
+    """Raises an AssertionError if two geometry array_like objects are not equal.
+
+    Given two array_like objects, check that the shape is equal and all elements of
+    these objects are equal. An exception is raised at shape mismatch or conflicting
```

Comment:
Currently this uses `equals_exact` for structural coordinate equality. Should we add an option to be able to get spatial equality (`equals`) as well?

Reply:
In addition, for equals_exact it might also be nice to add a `normalize=True` option (so you don't have to do that manually beforehand)